program-runtime/src/loaded_programs.rs

```diff
@@ -102,6 +102,32 @@ impl From<ProgramCacheEntryOwner> for Pubkey {
     }
 }
 
+/*
+    The possible ProgramCacheEntryType transitions:
+    DelayVisibility is special because in the cache it is stored as Loaded and
+    is only materialized temporarily by ProgramCacheForTxBatch::find().
+
+    Builtin re/deployment:
+    - Empty => Builtin in TransactionBatchProcessor::add_builtin
+    - Builtin => Builtin in TransactionBatchProcessor::add_builtin
+
+    Un/re/deployment (with delay and cooldown):
+    - Empty / Closed => DelayVisibility in UpgradeableLoaderInstruction::DeployWithMaxDataLen
+    - Loaded / FailedVerification => DelayVisibility in UpgradeableLoaderInstruction::Upgrade
+    - Loaded / FailedVerification => Closed in UpgradeableLoaderInstruction::Close
```

Comment:
Can Unloaded => Closed happen?

Reply:
In order to close a program you have to mention it in a Message which will cause it to be loaded first.
So it would be Unloaded => Loaded => Closed.

Reply:
I see makes sense